Vue3: fix source decorator + rendering with decorators

chakAs3 opened the pull request:
Closes #20670  Closes #21235  may even Closes #21221  stil have to go through some testing 



##  **Improving Reactivity and Rendering in Storybook**

In this PR i took a bit time to get things right to minimise issue in future, I'm excited to share some of the enhancements I've made to improve the reactivity and rendering of stories.

###  **Fully Reactive Stories**
One of the most significant improvements I've made is to ensure that all stories are fully reactive to related changes, without requiring a remount or re-rendering of the DOM tree. This is achieved by making the top-level components reactive via the props and even the slots, which fixes a lot of issues related to remounting components or the entire Vue app and causing the loss of state.
Furthermore, I've added support for reactive stories using multiple decorators. This is particularly helpful for developers who use UI libraries like Chakra or Vuetify to build fast apps or prototypes. These libraries need to wrap their stories to inject their style or configuration. Now, with CSF 2 and 3, we can work in a reactive way with decorators, which greatly simplifies the process.
To demonstrate these improvements, I've created a repository that showcases these points. You can find the link 
check out [this video demo](https://photos.app.goo.gl/boDAkqQ3rGsDyzWn8)
### **Powerful and Reactive Slots**
Another area where I've made significant improvements is in the handling of slots. Slots are now very powerful, as they are reactive and automatically generated if your component has defined them. We support named and scoped slots, and using the args, you can pass slot types such as function, string, VNode (using Vue render function h), or Functional Component. This is incredibly powerful, as it opens up a wide range of possibilities for building custom components and integrating them seamlessly into your storybook.
One area where we still have work to do is on the controls side, as we link between components using the controls and inject them as slots to any component. However, with the improvements we've made, we're confident that we can continue to push the boundaries of what's possible with slots in Storybook.
To see these improvements in action, check out [this video demo](https://photos.app.goo.gl/o43G72VYKkogt3Lz8).

### **Source Decorator**
Last but not least, I've rewrote my old Source Decorator . This is a fun and important feature that most developers will appreciate, as it makes it easy to pick working code and use it quickly. With the Vue compiler, we can get proper DOM elements to parse and generate the right and proper code.
Overall, I'm thrilled to have contributed to the Storybook code base and made these improvements. I believe that they will significantly improve the developer experience and enable developers to build more powerful and flexible components.
check out [this video demo](https://photos.app.goo.gl/9eakGoHJzdK8q5eU7).

@shilman @jonniebigodes  let me know if we can work on documentation, i will be working on short video for vue 3 developers showing new features in V7

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

chakAs3 committed:
rewrite source decorator with optimisation

chakAs3 committed:
some experimental

chakAs3 committed:
Merge branch 'next-remote' into chaks/vue3-source-fix

chakAs3 committed:
start writing  propre tests for source decorator

chakAs3 committed:
write tests for 2 main src decorator functions

chakAs3 committed:
add new tests

chakAs3 committed:
adjusting the render with tests

chakAs3 committed:
add more tests for sources decorator

chakAs3 committed:
testing the render

chakAs3 committed:
vue3 fixing prepareStory for decotators

chakAs3 committed:
some refactoring and cleanup

chakAs3 committed:
after merging next branch

chakAs3 committed:
fix null slots

chakAs3 committed:
handle no slots , null issue

chakAs3 committed:
fix args Inheritance and reactivity

chakAs3 committed:
merging next branch

chakAs3 committed:
fix null value exception for mdx stories

chakAs3 committed:
fix slot double quote display

chakAs3 committed:
fix source for mdx using originFn and context

chakAs3 committed:
refactory generateSource to be elegante

chakAs3 committed:
add some tests for source decorator and render

chakAs3 committed:
merging next to local branch

chakAs3 committed:
add some tests and fix CSF2 with decorator

chakAs3 committed:
fixing react decorator breaks the reactivity

chakAs3 committed:
complete merging next branch

chakAs3 committed:
use remount as easier way

chakAs3 committed:
emit using channel forceRemount

chakAs3 committed:
cleanup code

chakAs3 committed:
fix args after tests + cleanup

chakAs3 committed:
attribute style dynamic source

chakAs3 committed:
string double quote isse

chakAs3 committed:
v-bind , style display as string

chakAs3 committed:
style should be dynamic in source

chakAs3 committed:
cleanup after testing.

chakAs3 committed:
refactor add utils to docs

chakAs3 committed:
merging next branch

chakAs3 committed:
add utils

chakAs3 committed:
use vue watch() instead of useEffect()

chakAs3 committed:
merge next branch

chakAs3 committed:
just to rerun the ci, it did pass on server y??

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 committed:
fix slots reactivity without render function temp

chakAs3 committed:
fix  undefined attr + cleanup

chakAs3 committed:
merge next branch

chakAs3 committed:
run decorators in renderToCanvas save reactivity

chakAs3 committed:
great reasult finally

chakAs3 committed:
merge next branch

chakAs3 committed:
add try catch in case decorator are react Fn use

chakAs3 committed:
slots properly rendered , and source generated

chakAs3 committed:
cleanup

chakAs3 committed:
source generating for slots with recusivity with

kasperpeulen commented:
@chakAs3 Thanks 🙌  Will look at this monday!

chakAs3 commented:
> @chakAs3 Thanks 🙌 Will look at this monday!

sure man thx i know it is quite large PR to review .

chakAs3 committed:
refactory and cleanup

chakAs3 committed:
merge next branch

chakAs3 committed:
fix the units tests

chakAs3 committed:
remove console logs

chakAs3 committed:
cleanup oode and try to solve yn pkg issue

kasperpeulen commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -30,13 +32,25 @@ function prepare(
       return h(story);
     },
+    renderTracked(event) {
+      // console.log('story renderTracked', event); // this works only in dev mode
```
Does this need to be removed?

chakAs3 replied:
well we could do as i mentioned anyway it runs only in dev mode, it helps other contributor to debug if they need

kasperpeulen commented on code/lib/store/template/stories/globals.stories.ts:
```diff
@@ -29,6 +29,7 @@ export const Events = {
   ],
   play: async ({ canvasElement }: PlayFunctionContext<any>) => {
     const channel = globalThis.__STORYBOOK_ADDONS_CHANNEL__;
+    await channel.emit('updateGlobals', { globals: { foo: 'fooValue' } });
```
Why is this added?

chakAs3 replied:
the same raison as previous the discussed one time with @JReinhold  i think we need to place the reset state in the begining of the play function ,because if in case your play function fails in the middle will never reach the end to reset the state, so if you navigate back to this story you will not pass unless you do it manually.

i think it safer to the state in the begining of the play function

kasperpeulen commented on code/lib/store/template/stories/args.stories.ts:
```diff
@@ -73,6 +73,7 @@ export const Events = {
     await within(canvasElement).findByText(/initial/);
 
     await channel.emit(UPDATE_STORY_ARGS, { storyId: id, updatedArgs: { test: 'updated' } });
+    await new Promise((resolve) => channel.once(STORY_ARGS_UPDATED, resolve));
```
@tmeasday @ndelangen Can you look at this, I don't know much about this API. Does it makes sense to await this in every test for every renderer?

chakAs3 replied:
we need to normalise our code as in many place you wait for UPDATED event to resolve and other no.
i took the oportunity to do that.

kasperpeulen commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -64,8 +81,31 @@ export function decorateStory(
+  if (update) {
+    const { args, argTypes } = update;
+    if (args && !argTypes) {
+      const deepCopy = JSON.parse(JSON.stringify(args)); // avoid reference to args
```
What if the args are not primitives? Function vs classes?
Do we need the deep copy?

chakAs3 replied:
i think i will remove or change with 
```t
const argCopy = { ...args }
```
i have to test it again, the issue is if the updated args has reference to context.args which is most the time , in this case you will applying change on both of them, what i need is the values only.

kasperpeulen commented on code/renderers/vue3/src/render.ts:
```diff
@@ -22,65 +30,128 @@ export const setup = (fn: (app: any) => void) => {
+        const slotValue = context.args[key];
+        if (typeof slotValue === 'function' || isVNode(slotValue)) return h(slotValue);
+        if (Array.isArray(slotValue)) return slotValue.map((item: any) => h(item));
+        if (typeof slotValue === 'object') return JSON.stringify(slotValue);
```
I don't really understand why we would need all of this, in the type in TS we only allow for:

* A `VNodeChild`
* or a function that accept props and returns a `VNodeChild`

If it is a function we can just return the value right? As the h function accepts that?
If it is an array, it must be an array of VNodeChild, and the h function also accepts that for child.
If it is an object, it must be a VNode, why should we stringify that?

It this is about doing something sensible when the user gives incorrect args, I would rather throw when the slot arg is not a Vnode or a function.

chakAs3 replied:
Yes i got your point but dont forget the controls you can pass and object, rather we handle this case, for now we can just print , that why i stringify

chakAs3 replied:
So for me the JSON like a fallback in case the object is not a VNode, there is no harm to have it .
1- VNode
2- Function Component 
3- Array of VNode  we need this 100% very usefull generating in UI , like list item , or carousel ...

kasperpeulen replied:
Actually, this doesn't take into account slots accepting inner props anymore:
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/1035299/224734655-2b39b7b3-a92e-4a8f-8932-7156e9d94381.png">

I am not a fan of adding complexity here, it easily leads to more bugs. The h function is already powerful, I think the concern of storybook, should be to only pass the right args to slot function:
`h(Component, props, slot)`

And should not add any arbitrary mapping logic on top of the `h` function, IMO.

I think it is best if it works similar as if the user would write a custom render function based on `h`. For example:

```
export const Primary: Story = {
  render: () => {
    return h(
      BaseLayout,
      { otherProp: true },
      {
        header: ({ title }) =>
          h({
            components: { Button },
            template: `<Button :primary='true' label='${title}'></Button>`,
          }),
        default: [
          'some text',
          h(Button, { primary: true, label: 'default' }),
        ],
        footer: h(Button, { primary: true, label: 'footer' }),
      }
    );
  },
};
```

> 2- Function Component

I think the user should wrap themself the Function component in a `h` function.

> 3 Array of VNode we need this 100% very usefull generating in UI , like list item , or carousel ...

You can already pass array of VNode's to the slots in the `h` function, so we don't have to do anything there.

> 4 and you forgot the basic one which just a text or function returning text. before i go for complex type user may need just to pass a text

With `VNodeChild` I mean the following type def from the vue source:

```ts
type VNodeChild = VNodeChildAtom | VNodeArrayChildren;

type VNodeArrayChildren = Array<VNodeArrayChildren | VNodeChildAtom>;

type VNodeChildAtom = VNode | string | number | boolean | null | undefined | void;
```

So string, number boolean, null and undefined are covered there automatically by the `h` function. And arrays as well.

> Yes i got your point but dont forget the controls you can pass an object

This is also true for React, but we don't stringify children in React as well. Maybe at some point we could solve this more generally for all renderers.  For example, we should only allow the control of slots for any renderer to be updated with plain strings.

chakAs3 replied:
> And should not add any arbitrary mapping logic on top of the `h` function, IMO.

i totally agree, i look at it now it is like doing extra mapping, it would work simple like this 
```ts 
 return typeof slotValue === 'function' ? h(slotValue) : slotValue;
```

if we ignore the JSON, as we said earlier we need to work on the controls

chakAs3 replied:
> > And should not add any arbitrary mapping logic on top of the `h` function, IMO.
> 
> i totally agree, i look at it now it is like doing extra mapping, it would work simply like this
> 
> ```ts
>  return typeof slotValue === 'function' ? h(slotValue) : slotValue;
> ```
> 
> if we ignore the JSON, as we said earlier we need to work on the controls

kasperpeulen replied:
Actually, this still doesn't work with inner props, we probably should add a test case for that, as it easy to regress:

![image](https://user-images.githubusercontent.com/1035299/225042667-ec34980c-533e-41f6-91d6-e44a96bd3c5b.png)

See that my header is getting undefined, as the innerProps are not correctly passed.

It should be like this, to work correctly with innerProps:

```ts
  .map(([key, value]) => {
    return [key, context.args[key]];
  });
```

Or maybe better:

```ts
  .map(([key, value]) => {
    const slotValue = context.args[key];
    const normalized = typeof slotValue === 'function' ? slotValue : () => slotValue;
    return [key, normalized];
  });
```

As Vue gives performance warnings if slots are not specified as functions.

kasperpeulen commented:
@chakAs3 Thank you very much, these are really good improvements.
I'm not experienced with Vue, nor the core rendering logic of storybook, so I'm trying to wrap my head around it now.
I hope we can get this in before storybook day.

Do you know what is going on with the tests and build?

@ndelangen @tmeasday If any of you have time, it would be good if you can look at this as well.

chakAs3 committed:
cleanup and refactory

chakAs3 committed:
Merge remote-tracking branch 'origin/next' into chaks/vue3-source-fix

chakAs3 commented:
> @chakAs3 Thank you very much, these are really good improvements. I'm not experienced with Vue, nor the core rendering logic of storybook, so I'm trying to wrap my head around it now. I hope we can get this in before storybook day.
> 
> Do you know what is going on with the tests and build?
> 
> @ndelangen @tmeasday If any of you have time, it would be good if you can look at this as well.

yes @kasperpeulen  thank you for taking time and review, i will fix the tests i don't know why it is failing

chakAs3 committed:
fix prep, pass tests

chakAs3 committed:
Merge remote-tracking branch 'origin/next' into chaks/vue3-source-fix

chakAs3 committed:
refactory get slots

chakAs3 committed:
fix args updates and correct unit tests

chakAs3 committed:
fix args update

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

JReinhold commented:
@chakAs3 can you provide a demo or an explanation of how this works with slots, and what the API looks like from a user perspective? I want to understand if this is useful and possible for other slot-based languages as well, like Svelte.

tmeasday commented on code/renderers/vue3/src/render.ts:
```diff
@@ -22,65 +30,118 @@ export const setup = (fn: (app: any) => void) => {
+ */
 
-  return Object.fromEntries(slots);
+function updateContextDecorator(
+  storyFn: PartialStoryFn<VueRenderer>,
+  storyContext: StoryContext<VueRenderer>
+) {
+  const storyDecorators: Set<DecoratorFunction<VueRenderer>> = (
+    storyContext.hooks as HooksContext<VueRenderer>
+  ).mountedDecorators;
+
+  if (storyDecorators && storyDecorators.size > 0) {
+    storyDecorators.forEach((decorator: DecoratorFunction<VueRenderer>) => {
+      try {
+        if (typeof decorator === 'function') {
+          decorator((u) => {
+            if (u && u.args && !u.globals) return storyFn();
+            return () => {};
+          }, storyContext);
+        }
+      } catch (e) {
+        // in case the decorator throws an error, we need to re-render the story
+        // mostly because of react hooks that are not allowed to be called conditionally
+        reactiveState.globals = storyContext.globals; // { ...storyContext.globals };
+      }
+    });
+  }
 }
```
Can you explain what this is doing a little more? I'm a little puzzled about what's happening inside here.

chakAs3 replied:
yes i have  vue is fine grained reactive, so prepareStory runs 1 time to create the Ui, however we may have decorators where global or in Story level some alters the args,  like 
```ts
 (storyFn: PartialStoryFn, context: StoryContext) => {
      const { argNames } = context.parameters;
      const object = argNames ? pick(context.args, argNames) : context.args;
      return storyFn({ args: { object } });
    },
  ],
```

so i need to run these to update my reactive context.
in case of Decorators that have react hooks will through an error  that i catch to update the context.glabals , if there is the change i will rerender the story so the react hooks got executed

tmeasday replied:
So if I am understanding correctly, the issue is that if a decorator:

 - Reads one arg
 - Sets or overrides another arg

then without this code the decorator will not re-run when the first arg changes, so the second arg will have the incorrect value? And the reason for that is that the way the `renderToDom`/`applyDecorators` function is setup, means that when the first arg changes, the decorator would not otherwise re-run?

I think this isn't the right solution to that problem. Here are some other issues I can see with the above off the top of my head:

1. Storybook hooks won't work (you said "react hooks" but I assume you mean SB hooks). What if the decorator relies on them?

3. What if the decorator changes the actual rendered output? Will that get dropped? What if the decorator changes some other field?

I also am not sure about the `if (u && u.args && !u.globals)` guard. What's the idea there?

I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should re-run whenever args change. If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args). This approach seems fraught with problems.

chakAs3 replied:
Hi T

> So if I am understanding correctly, the issue is that if a decorator:
> 
> * Reads one arg
> * Sets or overrides another arg
> 
> then without this code the decorator will not re-run when the first arg changes, so the second arg will have the incorrect value? And the reason for that is that the way the `renderToDom`/`applyDecorators` function is setup, means that when the first arg changes, the decorator would not otherwise re-run?
> 
> I think this isn't the right solution to that problem. Here are some other issues I can see with the above off the top of my head:
> 
> 1. Storybook hooks won't work (you said "react hooks" but I assume you mean SB hooks). What if the decorator relies on them?
> 2. What if the decorator changes the actual rendered output? Will that get dropped? What if the decorator changes some other field on the context?
> 
> I also am not sure about the `if (u && u.args && !u.globals)` guard. What's the idea there?
> 
> I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should "properly" re-run whenever args change (treating the story's render function as the final decorator). If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args, or make a `useArgs` style API that forces Vue to re-render them). This approach seems fraught with problems.

Hi @tmeasday  i thought about your question "updated context should be passed to story?" and i managed to get it done, it helped removed a code i did not like cause it was just workout for me you can check my new code

chakAs3 replied:
> > I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should "properly" re-run whenever args change (treating the story's render function as the final decorator). If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args, or make a `useArgs` style API that forces Vue to re-render them). This approach seems fraught with problems.

Yep i agree that is why i needed your opinion. i could rely drop that workaround i have it done like you said just passing the the updated context,  this fix the vue decorators, and for decorators with hook, i'm watching context.globals is there is a change i trigger the vue app render  after recreating the VDom. ( react hooks works)

tmeasday commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -49,10 +51,13 @@ export function decorateStory(
           ...sanitizeStoryContextUpdate(update),
         });
+        if (update) updateReactiveContext(context, update);
         return story;
```
Shouldn't we pass the updated context into `story`?

chakAs3 replied:
it is passed already i guess but i need to update the reactive context ( used to track the state for vue ).
@tmeasday  i was little bit fighting to get how storybook rendering system works by myself, i guess i got 90% but sure you will guide me to  get 10% remaining .

kasperpeulen replied:
Have you seen the Svelte implementation? @chakAs3 
We were thinking that the proper way for Vue maybe similar.
We might need to treat every decorator as an actual vue component.

cc @JReinhold

tmeasday commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -64,8 +69,31 @@ export function decorateStory(
+) {
+  if (update) {
+    const { args, argTypes } = update;
+    if (args && !argTypes) {
```
What happens when `argTypes` is changed? (I don't know if this should be allowed anyway, by default we disallow `argTypes` updates: https://github.com/storybookjs/storybook/blob/2345f208aaa4477b9ce3f79e8942e4acfd7a00f2/code/lib/preview-api/src/modules/store/decorators.ts#L39

tmeasday replied:
Is it possible to compare the args to the previous values perhaps?

tmeasday commented:
Maybe we could add some template stories to the renderer demonstrating the new apis/behaviours also

chakAs3 commented:
> Maybe we could add some template stories to the renderer demonstrating the new apis/behaviours also

i have already 1 repo planing outside of the monorepo, planning run some tests there so i don't impact the monorepo in case other framework, otherwise yes i will add some stories there

chakAs3 commented:
> @chakAs3 can you provide a demo or an explanation of how this works with slots, and what the API looks like from a user perspective? I want to understand if this is useful and possible for other slot-based languages as well, like Svelte.

yes sure i already prepare videos and 1 repo

tmeasday commented:
@chakAs3 if you add story files in the renderer (like these ones: https://github.com/storybookjs/storybook/blob/01e21c9470f7eeb395405e30252dc0c40321100c/code/renderers/react/template/stories/hooks.stories.tsx) they'll get tested by the test runner + chromatic which gives us much better test coverage of this new behaviour!

chakAs3 commented:
> @chakAs3 if you add story files in the renderer (like these React ones: https://github.com/storybookjs/storybook/blob/01e21c9470f7eeb395405e30252dc0c40321100c/code/renderers/react/template/stories/hooks.stories.tsx) they'll get tested by the test runner + chromatic which gives us much better test coverage of this new behaviour!

yes tom i have already added the reactiveArgs stories 2 months ago , i will add these some others

kasperpeulen commented:
Hi @chakAs3. There is a lot going on in this PR.

Could we split this in more focused PRs. We are currently in a feature freeze for Storybook 7. 
So we only want to merge bugs now, and release new features shortly after the Storybook 7 release in a SB 7.1 alpha release.

Could we split this PR maybe in 3 different PRs:
* Reactive stories and decorators
* Source decorators
* Reactive slots 

I'm also not 100% sure, which parts are fixing bugs, and which are introducing new features.

Thanks for all your work!

chakAs3 committed:
fix hooks story tests, by triggering render all tree

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 committed:
add slots test story to vue3 renderer templates

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 commented:
> Hi @chakAs3. There is a lot going on in this PR.
> 
> Could we split this in more focused PRs. We are currently in a feature freeze for Storybook 7. So we only want to merge bugs now, and release new features shortly after the Storybook 7 release in a SB 7.1 alpha release.
> 
> Could we split this PR maybe in 3 different PRs:
> 
> * Reactive stories and decorators
> * Source decorators
> * Reactive slots
> 
> I'm also not 100% sure, which parts are fixing bugs, and which are introducing new features.
> 
> Thanks for all your work!

hi @kasperpeulen thanks for quality of your review. you really took time to do that in good manner and helped me changing some code.  

i think spliting in this stage is difficult i can't ship code that i'm sure 100% will have a bug,i know exactly what will happen.

i have add tests story for slots like @tmeasday asked and answered his questions

chakAs3 committed:
using storyFn to update args from decorators

chakAs3 committed:
get it done properly yeeeey

chakAs3 committed:
refactory after testings

chakAs3 committed:
fix tests

kasperpeulen commented on code/renderers/vue3/template/stories/ReactiveSlots.stories.js:
```diff
@@ -0,0 +1,49 @@
+export const ReactiveSlotTest = {
+  args: {
+    default: 'Default Text Slot',
+    header: () => h('h1', 'Header Functional Component Slot'),
```
Could we change this so that the title inner prop is used?

`header: ({title}) => h('h1', title),`

Would be great if we can test that this doesn't regress.

kasperpeulen commented on code/renderers/vue3/src/render.ts:
```diff
@@ -22,65 +23,86 @@ export const setup = (fn: (app: any) => void) => {
-    .map(([key, value]) => [key, typeof value === 'function' ? value : () => value]);
-
-  return Object.fromEntries(slots);
+    .map(([key, value]) => [
```
This has be to changed to:

```ts
  .map(([key, value]) => {
    const slotValue = context.args[key];
    return [key, typeof slotValue === 'function' ? slotValue : () => slotValue];
  });
```

To make sure innerProps are not ignored.

tmeasday commented on code/lib/core-server/src/dev-server.ts:
```diff
@@ -48,6 +49,7 @@ export async function storybookDevServer(options: Options) {
   }
 
   app.use(getAccessControlMiddleware(core?.crossOriginIsolated));
+  app.use(getCachingMiddleware());
```
What's this part about? Could it be a separate PR?

tmeasday commented on code/renderers/vue3/template/stories/ReactiveSlots.stories.js:
👏

chakAs3 committed:
add scoped slot tests stories

kasperpeulen commented on code/lib/core-server/src/utils/get-caching-middleware.ts:
```diff
@@ -0,0 +1,8 @@
+
+export function getCachingMiddleware(): RequestHandler {
+  return (req, res, next) => {
+    res.header('Cache-Control', 'no-store');
```
@ndelangen Didn't you did this as well?

kasperpeulen commented on code/renderers/vue3/template/stories/ScopedSlots.stories.js:
```diff
@@ -0,0 +1,98 @@
+
+export const ReactiveScopedSlot = {
+  args: { label: 'Storybook Day', year: 2022 },
+  render: (args) => ({
```
We should have a test for not using a custom render function as well:

```ts
  args: {
    label: 'Storybook Day',
    year: 2022,
    default: ({ text, year }) => `${text}, ${year}`,
  },
```

but this crashes in the source decorator, on line 92:

<img width="945" alt="image" src="https://user-images.githubusercontent.com/1035299/225331990-1346b1e1-c40e-4adc-a5e3-01a03a837f10.png">

kasperpeulen replied:
@chakAs3 This is supported by `h` function of Vue, so we should support this as well. You don't need a custom render function for this.

kasperpeulen commented on code/renderers/vue3/src/docs/sourceDecorator.ts:
```diff
@@ -36,89 +43,80 @@ const skipSourceRender = (context: StoryContext<Renderer>) => {
+    let slotContent = '';
+
+    if (typeof slot === 'function') {
+      const slotVNode = slot();
```
This line crashes when using scoped slots without render function:

<img width="1695" alt="image" src="https://user-images.githubusercontent.com/1035299/225333243-bb9717c0-a053-4a02-8327-af1374bbe3cf.png">

```
TypeError: Cannot destructure property 'text' of 'undefined' as it is undefined.
    at default (ScopedSlots.stories.js:15:7)
    at config.mjs:7671:25
    at Array.map (<anonymous>)
    at mapSlots (config.mjs:7667:34)
    at generateComponentSource (config.mjs:7745:81)
    at config.mjs:7753:61
    at Array.map (<anonymous>)
    at generateTemplateSource (config.mjs:7753:36)
    at generateSource (config.mjs:7777:29)
    at sourceDecorator (config.mjs:7759:3)
```

If I put a default value, it doesn't crash:

<img width="1689" alt="image" src="https://user-images.githubusercontent.com/1035299/225333623-e6434c26-1e10-45c3-9a33-8ae7cd6d393c.png">

kasperpeulen replied:
@chakAs3 I updated the test a bit. It work now in canvas mode, but still break in docs mode:

<img width="1070" alt="image" src="https://user-images.githubusercontent.com/1035299/225616707-6c02de40-c8f5-4641-898c-9d5a57459aa6.png">

kasperpeulen replied:
It seems that only templates are tested there, but not if you use the default render function.

chakAs3 committed:
added arg to Reactive SLot test story

chakAs3 committed:
add test story without render fn for Scoped slots

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 committed:
Merge branch 'next' into chaksfi/vue3-source-fix

chakAs3 committed:
get back eval for printing the object value

chakAs3 committed:
i don't know this passed the tests and now does not, some on reat:check wierd

chakAs3 reviewed (commented):
yes i can see that i working on it

chakAs3 committed:
remove files

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

kasperpeulen committed:
Improve ScopedSlot test and TS test coverage

kasperpeulen committed:
Add caching middleware back

kasperpeulen commented:
@chakAs3 Right I thought maybe about a test like:

1. Filling in an input element
2. Changing args
3. Input state should be preserved

kasperpeulen commented on code/renderers/vue3/src/render.ts:
```diff
@@ -22,65 +23,81 @@ export const setup = (fn: (app: any) => void) => {
+  // if the story is already rendered and we are not forcing a remount, we just update the reactive args
   if (existingApp && !forceRemount) {
-    updateArgs(existingApp.reactiveArgs, reactiveArgs);
+    const element = storyFn();
```
Ideally, I think we should only call storyFn during `setup`, not during updating or render?

chakAs3 replied:
Yeap that is exactly what i'm talking about, but because the decorator system and how  Storybook is handling the render is using React concept which is not reactive you have to run all the stack in every change

kasperpeulen replied:
What, if we would get rid of hooks in all the default decorators in storybook, and explain in the docs that hooks can not be used in user based vue decorators, would that solve the problem?

chakAs3 replied:
Yeah but these hooks are used by Storybook, no Vue developer will use hooks.
Actually Storybook creating 7 layers of decorators which are basically React Components (  Functional components )
that wrap the story whether is vue,preact, or any other renderer, these decorator are responsible of background, theme 
![image](https://user-images.githubusercontent.com/711292/226339911-3c218732-3041-4d6c-8969-a67f4aea853f.png)
to update the DOM, react needs story renderer ( vue, preact, angular ) to re-render all the tree in order for the hooks to update the DOM

kasperpeulen commented on code/renderers/vue3/src/render.ts:
```diff
@@ -22,65 +23,81 @@ export const setup = (fn: (app: any) => void) => {
+        () => reactiveState.globals,
+        () => {
+          storyContext.globals = reactiveState.globals;
+          rootElement = storyFn();
```
Ideally, we should try a solution where is not needed right?

chakAs3 replied:
Yes But for now Tom is saying we stick with react until we found a solution.

chakAs3 committed:
slots test story +Test reactivity with decorator and CSF2

chakAs3 committed:
cleaup logs

chakAs3 committed:
add utility function

chakAs3 committed:
Merge branch 'chaks/vue3-source-fix' of https://github.com/storybookjs/storybook into chaks/vue3-source-fix

chakAs3 committed:
add Slot Story with Render to test interpolation & fixing unit test

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 commented on code/renderers/vue3/template/stories/ReactiveArgs.stories.js:
```diff
@@ -42,3 +42,86 @@ export const ReactiveTest = {
+};
+
+// to test that Simple html Decorators in CSF2 format are applied correctly in reactive mode
+export const ReactiveCSF2Wrapper = (args, { argTypes }) => ({
```
these 2 Stories to show the issue on the next branch, Stories are not reactive they decorated, whether CSF3 or CSF2

chakAs3 committed:
fix variable declaration source

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 committed:
fix unit tests

chakAs3 committed:
remove logs..

chakAs3 committed:
fix TS issue , check not passing

chakAs3 committed:
Merge branch 'next' into chaks/vue3-soe-fix

chakAs3 commented:
> Right, I noticed that as well. My feeling that the correct way forward would be to rewrite them all without hooks, or at least something that is compatible with signal based architectures. (Initial setup only running once).

@JReinhold  

yes I though of having something like preact + signals will realy do great in performance,  when i was the hooks i though these are react but Tom explained to me these are Storybook hooks, but i guess they work pretty much like React.

in the source decorator i replace them with Vue watch

kasperpeulen commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -46,25 +47,31 @@ export function decorateStory(
       }
-
-      return prepare(decoratedStory, h(story, context.args)) as VueRenderer['storyResult'];
+      const props = story.props ?? context.args;
```
Why do we prefer story.props above context.args? This only works when `story` is a VNode right? And not component options.

chakAs3 replied:
here we could use just context.args, but we need to check first if the story args has been changed by some decorator, in these case inner story props and context.args are not same.

in your story in you added a decorator like this 
export const InheritanceStory = {
  decorators: [
    (storyFn: PartialStoryFn, context: StoryContext) =>
      storyFn({ args: { ...context.args, text: `story ${context.args['text']}` } }),
  ],
  args: {
    text: 'starting',
  },
}
```

This only works when story is VNode exactly  cause you had to pick the  args changes from the { update } decorator function, and pass them to the normal story ( not yet VNode )

chakAs3 committed:
fix ci add vite 4.2.0 to resolutions sandboxes

chakAs3 committed:
Merge branch 'next' into chaks/vue3-source-fix

chakAs3 commented on code/renderers/vue3/src/decorateStory.ts:
```diff
@@ -46,25 +47,31 @@ export function decorateStory(
-
-      return prepare(decoratedStory, h(story, context.args)) as VueRenderer['storyResult'];
+      const props = story.props ?? context.args;
+      const innerStory = () => h(story, props);
```
here the inner Story should be a function to be binded with the props changes, this makes CSF2  format works too

dxb-story committed:
Merge branch 'next' into chaks/vue3-source-fix

dxb-story committed:
Merge branch 'next' into chaks/vue3-source-fix

dxb-story committed:
remove the watch on context.globals, we call anyway storyFn on renderToCanvas

dxb-story committed:
merge next chaks/vue3-source-fix

chakAs3 committed:
merge next intot chaks/vue3-source-fix

blowsie commented:
Fwiw, @chakAs3 has [mentioned in discord](https://discord.com/channels/486522875931656193/1090204816300593162) that this should fix an issue with the templates, 

> Simply having multiple `<div></div>` blocks in the template : `` would cause this. (this happens if you have the same html elements within one another).


![image](https://user-images.githubusercontent.com/308572/229755729-5b354e76-4b20-48b6-a18a-2d1bd12b1268.png)

Reproduction here
https://discord.com/channels/486522875931656193/1090204816300593162/1091732602802679848
